Documentation for named tensors

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27173 Documentation for named tensors**

`docs/source/named_tensor.rst` is the entry point; most users will land
either here or the named tensor tutorial when looking to use named
tensors. We should strive to make this as readable, concise, and understandable
as possible.

`docs/source/name_inference.rst` lists all of the name inference rules.
It should be clear but it's hard to make it concise.

Please let me know if anything doesn't make sense and please propose
alternative wordings and/or restructuring to improve the documentation.
This should ultimately get cherry-picked into the 1.3 branch as one
monolithic commit so it would be good to get all necessary changes made
in this PR and not have any follow ups.

Test Plan:
- built and reviewed locally with `cd docs/ && make html`.

Differential Revision: [D17763046](https://our.internmc.facebook.com/intern/diff/D17763046)

zou3519 committed:
Documentation for named tensors

`docs/source/named_tensor.rst` is the entry point; most users will land
either here or the named tensor tutorial when looking to use named
tensors. We should strive to make this as readable, concise, and understandable
as possible.

`docs/source/name_inference.rst` lists all of the name inference rules.
It should be clear but it's hard to make it concise.

Please let me know if anything doesn't make sense and please propose
alternative wordings and/or restructuring to improve the documentation.
This should ultimately get cherry-picked into the 1.3 branch as one
monolithic commit so it would be good to get all necessary changes made
in this PR and not have any follow ups.

Test Plan:
- built and reviewed locally with `cd docs/ && make html`.

nairbv commented on docs/source/name_inference.rst:
```diff
@@ -0,0 +1,426 @@
+
+Named tensors use names to check that some APIs are being used correctly and
+propagate names in a process called name inference.
+Name inference works as the following:
```
maybe "does the following" ? "works as the following" sounds odd to me.

gchanan replied:
works as follows?

zou3519 replied:
I'm not the best writer, please keep sending comments of things like this that can be improved.

nairbv commented on docs/source/name_inference.rst:
```diff
@@ -0,0 +1,426 @@
+and their name inference rules. If an operation is not here then it is probably not
+supported but one can try to run the operation and see what happens.
+
+.. csv-table:: Supported Operations
```
is there any way this table can be generated or can refer to some other list? I think people may forget to add operators as we build out support, leaving this incorrect.

zou3519 replied:
It's generated from a csv file that I keep locally. In the long term view of things we will get rid of the table as more operations support named tensors and directly inline name propagation rules into the API references for each operator. 

In the short term, I agree it would be nice to auto generate this table from some source, maybe even combining it with named tensor testing in `test_namedtensor.py`. That task is beyond the scope of this PR, though and would require some solid legwork.

nairbv commented on docs/source/named_tensor.rst:
```diff
@@ -0,0 +1,279 @@
+   .. automethod:: align_to
+
+   .. automethod:: unflatten
+   .. py:method:: flatten(dims, out_dim) -> Tensor
```
I don't know rst well, but the git rendering of the diff seems to suggest this needs an escape character

zou3519 replied:
I'll play around with it but I'm not convince that the git rendering is correct. The documentation builds and looks correct locally (ignore the font):
![image](https://user-images.githubusercontent.com/5652049/66062621-bfffe700-e50f-11e9-8fd1-5f52413fb7eb.png)

zou3519 committed:
Update on "Documentation for named tensors"


`docs/source/named_tensor.rst` is the entry point; most users will land
either here or the named tensor tutorial when looking to use named
tensors. We should strive to make this as readable, concise, and understandable
as possible.

`docs/source/name_inference.rst` lists all of the name inference rules.
It should be clear but it's hard to make it concise.

Please let me know if anything doesn't make sense and please propose
alternative wordings and/or restructuring to improve the documentation.
This should ultimately get cherry-picked into the 1.3 branch as one
monolithic commit so it would be good to get all necessary changes made
in this PR and not have any follow ups.

Test Plan:
- built and reviewed locally with `cd docs/ && make html`.

gchanan commented on docs/source/name_inference.rst:
```diff
@@ -0,0 +1,426 @@
+propagate names in a process called name inference.
+Name inference works as the following:
+
+- Check names: an operator may check that certain dimensions must match.
```
certain dimension names must match?

gchanan commented on docs/source/name_inference.rst:
```diff
@@ -0,0 +1,426 @@
+- Propagate names: name inference computes and propagates names to output tensors.
+
+Below is a list of all operations that are supported with named tensors
+and their name inference rules. If an operation is not here then it is probably not
```
their associated name inference rules.

zou3519 commented on docs/source/name_inference.rst:
```diff
@@ -0,0 +1,426 @@
+.. warning::
+    The named tensor API is experimental and subject to change.
+
+Named tensors use names to check that some APIs are being used correctly and
```
My plan for this page is that users should read `docs/source/named_tensor.rst` first. This page is more of a long reference. Given that, I can add a link to docs/source/named_tensor.rst here -- do you still think that we should have a better opener here?

zou3519 replied:
I really like your introduction, though, so I'm going to add it to the other page (docs/source/named_tensor.rst)

zou3519 committed:
Update on "Documentation for named tensors"


`docs/source/named_tensor.rst` is the entry point; most users will land
either here or the named tensor tutorial when looking to use named
tensors. We should strive to make this as readable, concise, and understandable
as possible.

`docs/source/name_inference.rst` lists all of the name inference rules.
It should be clear but it's hard to make it concise.

Please let me know if anything doesn't make sense and please propose
alternative wordings and/or restructuring to improve the documentation.
This should ultimately get cherry-picked into the 1.3 branch as one
monolithic commit so it would be good to get all necessary changes made
in this PR and not have any follow ups.

Test Plan:
- built and reviewed locally with `cd docs/ && make html`.

zou3519 committed:
Update on "Documentation for named tensors"


`docs/source/named_tensor.rst` is the entry point; most users will land
either here or the named tensor tutorial when looking to use named
tensors. We should strive to make this as readable, concise, and understandable
as possible.

`docs/source/name_inference.rst` lists all of the name inference rules.
It should be clear but it's hard to make it concise.

Please let me know if anything doesn't make sense and please propose
alternative wordings and/or restructuring to improve the documentation.
This should ultimately get cherry-picked into the 1.3 branch as one
monolithic commit so it would be good to get all necessary changes made
in this PR and not have any follow ups.

Test Plan:
- built and reviewed locally with `cd docs/ && make html`.

zou3519 committed:
Update on "Documentation for named tensors"


`docs/source/named_tensor.rst` is the entry point; most users will land
either here or the named tensor tutorial when looking to use named
tensors. We should strive to make this as readable, concise, and understandable
as possible.

`docs/source/name_inference.rst` lists all of the name inference rules.
It should be clear but it's hard to make it concise.

Please let me know if anything doesn't make sense and please propose
alternative wordings and/or restructuring to improve the documentation.
This should ultimately get cherry-picked into the 1.3 branch as one
monolithic commit so it would be good to get all necessary changes made
in this PR and not have any follow ups.

Test Plan:
- built and reviewed locally with `cd docs/ && make html`.

zou3519 committed:
Update on "Documentation for named tensors"


`docs/source/named_tensor.rst` is the entry point; most users will land
either here or the named tensor tutorial when looking to use named
tensors. We should strive to make this as readable, concise, and understandable
as possible.

`docs/source/name_inference.rst` lists all of the name inference rules.
It should be clear but it's hard to make it concise.

Please let me know if anything doesn't make sense and please propose
alternative wordings and/or restructuring to improve the documentation.
This should ultimately get cherry-picked into the 1.3 branch as one
monolithic commit so it would be good to get all necessary changes made
in this PR and not have any follow ups.

Test Plan:
- built and reviewed locally with `cd docs/ && make html`.

gchanan commented on docs/source/name_inference.rst:
```diff
@@ -0,0 +1,428 @@
+
+Please read :ref:`named_tensors-doc` first for an introduction to named tensors.
+
+This document is a reference for *name inference*, a process for how named tensors
```
I think numbering here would make things clearer.  Something like..."a process that defines how named tensors:
1.  use names to provide additional automatic runtime correctness checks
2.  propagate names from input tensors to output tensors
"
(for symmetry you want to say either "from inputs to outputs" or "from input tensors to output tensors".

gchanan commented on docs/source/name_inference.rst:
```diff
@@ -0,0 +1,428 @@
+use names to automatically check that APIs are being called correctly at runtime and
+for how operators propagate names from input tensors to outputs.
+
+More formally, name inference consists of the following two steps:
```
I don't think this restatement is helpful.  You could just put the dimension matching as a specific correctness check above.

gchanan commented on docs/source/name_inference.rst:
```diff
@@ -0,0 +1,428 @@
+- Propagate names: name inference computes and propagates names to output tensors.
+
+Below is a list of all operations that are supported with named tensors
+and their associated name inference rules. If an operation is not here then it
```
This sentence isn't really actionable -- I can always try an operation and see what happens, you don't have to tell me that :).

You could say something like "If you don't see an operation listed here, but it would help your use case, please search if an issue has already been filed and if not file one"...

gchanan commented on docs/source/name_inference.rst:
```diff
@@ -0,0 +1,428 @@
+
+.. _removes_dimensions-doc:
+
+Removes dimensions
```
why do you call this removes dimensions instead of something with reduction in its name?

zou3519 replied:
Not everything that removes dimensions can be called a reduction. I.e., `tensor.select(0, ...)` is not always thought of as a reduction; only `sum, max, mean` etc are.

gchanan reviewed (changes requested):
is there a way to see the rendered doc?

zou3519 commented:
> is there a way to see the rendered doc?

You can check out the PR on local machine and build the docs locally. Other than that, we don't have an automated way of viewing pytorch/pytorch changes to documentation (we do have a way of previewing changes to pytorch/pytorch.github.io). I could host the rendered doc somewhere if I figure that out...

Instructions for how to see the rendered doc:
```
git fetch origin pull/27173/head:namedtensor_docs
git checkout namedtensor_docs
# build pytorch
cd docs
make html
open build/html
```

zou3519 commented:
@gchanan preview of the docs: http://auspicious-hand.surge.sh/named_tensor.html

zou3519 committed:
Update on "Documentation for named tensors"


`docs/source/named_tensor.rst` is the entry point; most users will land
either here or the named tensor tutorial when looking to use named
tensors. We should strive to make this as readable, concise, and understandable
as possible.

`docs/source/name_inference.rst` lists all of the name inference rules.
It should be clear but it's hard to make it concise.

Please let me know if anything doesn't make sense and please propose
alternative wordings and/or restructuring to improve the documentation.
This should ultimately get cherry-picked into the 1.3 branch as one
monolithic commit so it would be good to get all necessary changes made
in this PR and not have any follow ups.

Test Plan:
- built and reviewed locally with `cd docs/ && make html`.

zou3519 committed:
Update on "Documentation for named tensors"


`docs/source/named_tensor.rst` is the entry point; most users will land
either here or the named tensor tutorial when looking to use named
tensors. We should strive to make this as readable, concise, and understandable
as possible.

`docs/source/name_inference.rst` lists all of the name inference rules.
It should be clear but it's hard to make it concise.

Please let me know if anything doesn't make sense and please propose
alternative wordings and/or restructuring to improve the documentation.
This should ultimately get cherry-picked into the 1.3 branch as one
monolithic commit so it would be good to get all necessary changes made
in this PR and not have any follow ups.

Test Plan:
- built and reviewed locally with `cd docs/ && make html`.

gchanan commented on docs/source/named_tensor.rst:
```diff
@@ -0,0 +1,302 @@
+Named Tensors aim to make tensors easier to use by allowing users to associate
+explicit names with tensor dimensions. In most cases, operations that take
+dimension parameters will accept dimension names, avoiding the need to track
+dimensions by index. In addition, named tensors use names to automatically
```
nit: you probably want to say "position" here.  "dimensions by position" or "dimensions by index position" sounds good.

gchanan commented on docs/source/named_tensor.rst:
```diff
@@ -0,0 +1,302 @@
+dimension parameters will accept dimension names, avoiding the need to track
+dimensions by index. In addition, named tensors use names to automatically
+check that APIs are being used correctly at runtime, providing extra safety.
+Names can also be used to rearrange dimensions and to align tensors by their
```
you haven't defined what "align tensors by their position names mean.  Maybe something like:
"Names can also be used to rearrange dimensions, for example to support "broadcasting by name" rather than "broadcasting by position".

gchanan commented on docs/source/named_tensor.rst:
```diff
@@ -0,0 +1,302 @@
+Creating named tensors
+----------------------
+
+Factory functions now take a new names argument that represents a name for each dimension.
```
`names` should be coded.

gchanan replied:
"represents a name for each dimension" isn't quite accurate.  Maybe "associates a name with each dimension".

gchanan commented on docs/source/named_tensor.rst:
```diff
@@ -0,0 +1,302 @@
+    tensor([[0., 0., 0.],
+            [0., 0., 0.]], names=('N', 'C'))
+
+Named dimensions, like regular Tensor dimensions, are ordered and inherently positional.
```
you haven't defined what "inherently positional" means.  I'd just get rid of it.

gchanan commented on docs/source/named_tensor.rst:
```diff
@@ -0,0 +1,302 @@
+Named dimensions
+----------------
+
+See :attr:`~Tensor.names` for restrictions on tensor names. The most important
```
what makes the restriction important?  In general I think it's a good idea to avoid value-judgement words like this, that's up to the reader to decide.

gchanan commented on docs/source/named_tensor.rst:
```diff
@@ -0,0 +1,302 @@
+See :attr:`~Tensor.names` for restrictions on tensor names. The most important
+restriction is that tensors may not have two named dimensions with the same name.
+
+Use :attr:`~Tensor.names` to access the dimension names of a tensor and
```
show examples.

gchanan commented on docs/source/named_tensor.rst:
```diff
@@ -0,0 +1,302 @@
+Use :attr:`~Tensor.names` to access the dimension names of a tensor and
+:meth:`~Tensor.rename` to rename named dimensions.
+
+Named tensors coexist with unnamed tensors; named tensors are instances of
```
I'd say "can coexist" -- neither named tensors nor unnamed tensors exist until you create them.

zou3519 committed:
Update on "Documentation for named tensors"


`docs/source/named_tensor.rst` is the entry point; most users will land
either here or the named tensor tutorial when looking to use named
tensors. We should strive to make this as readable, concise, and understandable
as possible.

`docs/source/name_inference.rst` lists all of the name inference rules.
It should be clear but it's hard to make it concise.

Please let me know if anything doesn't make sense and please propose
alternative wordings and/or restructuring to improve the documentation.
This should ultimately get cherry-picked into the 1.3 branch as one
monolithic commit so it would be good to get all necessary changes made
in this PR and not have any follow ups.

Test Plan:
- built and reviewed locally with `cd docs/ && make html`.

Differential Revision: [D17763046](https://our.internmc.facebook.com/intern/diff/D17763046)

zou3519 commented:
Addressed comments and re-pushed to auspicious-hand.surge.sh

zou3519 committed:
Update on "Documentation for named tensors"


`docs/source/named_tensor.rst` is the entry point; most users will land
either here or the named tensor tutorial when looking to use named
tensors. We should strive to make this as readable, concise, and understandable
as possible.

`docs/source/name_inference.rst` lists all of the name inference rules.
It should be clear but it's hard to make it concise.

Please let me know if anything doesn't make sense and please propose
alternative wordings and/or restructuring to improve the documentation.
This should ultimately get cherry-picked into the 1.3 branch as one
monolithic commit so it would be good to get all necessary changes made
in this PR and not have any follow ups.

Test Plan:
- built and reviewed locally with `cd docs/ && make html`.

Differential Revision: [D17763046](https://our.internmc.facebook.com/intern/diff/D17763046)

zou3519 committed:
Update on "Documentation for named tensors"


`docs/source/named_tensor.rst` is the entry point; most users will land
either here or the named tensor tutorial when looking to use named
tensors. We should strive to make this as readable, concise, and understandable
as possible.

`docs/source/name_inference.rst` lists all of the name inference rules.
It should be clear but it's hard to make it concise.

Please let me know if anything doesn't make sense and please propose
alternative wordings and/or restructuring to improve the documentation.
This should ultimately get cherry-picked into the 1.3 branch as one
monolithic commit so it would be good to get all necessary changes made
in this PR and not have any follow ups.

Test Plan:
- built and reviewed locally with `cd docs/ && make html`.

Differential Revision: [D17763046](https://our.internmc.facebook.com/intern/diff/D17763046)

zou3519 committed:
Update on "Documentation for named tensors"


`docs/source/named_tensor.rst` is the entry point; most users will land
either here or the named tensor tutorial when looking to use named
tensors. We should strive to make this as readable, concise, and understandable
as possible.

`docs/source/name_inference.rst` lists all of the name inference rules.
It should be clear but it's hard to make it concise.

Please let me know if anything doesn't make sense and please propose
alternative wordings and/or restructuring to improve the documentation.
This should ultimately get cherry-picked into the 1.3 branch as one
monolithic commit so it would be good to get all necessary changes made
in this PR and not have any follow ups.

Test Plan:
- built and reviewed locally with `cd docs/ && make html`.

Differential Revision: [D17763046](https://our.internmc.facebook.com/intern/diff/D17763046)

zou3519 committed:
Update on "Documentation for named tensors"


`docs/source/named_tensor.rst` is the entry point; most users will land
either here or the named tensor tutorial when looking to use named
tensors. We should strive to make this as readable, concise, and understandable
as possible.

`docs/source/name_inference.rst` lists all of the name inference rules.
It should be clear but it's hard to make it concise.

Please let me know if anything doesn't make sense and please propose
alternative wordings and/or restructuring to improve the documentation.
This should ultimately get cherry-picked into the 1.3 branch as one
monolithic commit so it would be good to get all necessary changes made
in this PR and not have any follow ups.

Test Plan:
- built and reviewed locally with `cd docs/ && make html`.

Differential Revision: [D17763046](https://our.internmc.facebook.com/intern/diff/D17763046)
